File: hashes/0290-word-pattern.py
from itertools import zip_longest
import unittest

# https://leetcode.com/problems/word-pattern/

# python3 -m unittest hashes/0290-word-pattern.py


class Solution(unittest.TestCase):
    def wordPattern(self, pattern: str, s: str) -> bool:
        # The counts of distinct elements in two groups
        # and the count of distinct mappings are all equal.
        s = s.split(' ')
        return len(set(s)) == len(set(pattern)) == len(set(zip_longest(pattern, s)))

    def test(self):
        for pattern, s, expected in [
            ("abba", "dog cat cat dog", True),
            ("abba", "dog cat cat fish", False),
            ("aaaa", "dog cat cat dog", False),
        ]:
            output = self.wordPattern(pattern, s)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: hashes/0290-word-pattern.go
package hashes

import (
	"strings"
)

// https://leetcode.com/problems/word-pattern/

func wordPattern(pattern string, s string) bool {
	words := strings.Split(s, " ")

	n := len(words)
	if n != len(pattern) {
		return false
	}

	codes := make(map[string]byte)

	var used [26]bool

	var letter byte

	var ok bool

	var code int

	for idx := range n {
		if letter, ok = codes[words[idx]]; !ok {
			code = int(pattern[idx] - 'a')
			if used[code] {
				return false
			}

			used[code] = true
			codes[words[idx]] = pattern[idx]
		} else if letter != pattern[idx] {
			return false
		}
	}

	return true
}
